File: ctbrpgsca/costume.cpp
/*
Copyright (C) AD 2013-2020 Claudiu-Stefan Costea

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
*/
#include "costume.h"
#include "ability.h"
#include "actor.h"
#include "state.h"
#include "scene.h"
#include "role.h"

#include <QStringBuilder>

using namespace tbrpgsca;

QString Costume::CausesTxt = ", %1 is affected by %2";

int Costume::getAttack() const
{
    return this->atk;
}

int Costume::getDefense() const
{
    return this->def;
}

int Costume::getSpirit() const
{
    return this->spi;
}

int Costume::getWisdom() const
{
    return this->wis;
}

int Costume::getAgility() const
{
    return this->agi;
}

int Costume::getMaximumActions() const
{
    return this->mActions;
}

Ability& Costume::getAddedSkill(int const n) const
{
    return *(this->aSkills->at(n));
}

bool Costume::hasAddedSkill(Ability& skill) const
{
    QVector<Ability*>* aSkills = this->aSkills;
    return aSkills != nullptr && aSkills->contains(&skill);
}

int Costume::getAddedSkillsSize() const
{
    QVector<Ability*>* aSkills = this->aSkills;
    return aSkills == nullptr ? 0 : aSkills->size();
}

Ability& Costume::getCounterSkill(int const n) const
{
    return *(this->counters->at(n));
}

bool Costume::hasCounterSkill(Ability& skill) const
{
    QVector<Ability*>* counters = this->counters;
    return counters != nullptr && counters->contains(&skill);
}

int Costume::getCounterSkillsSize() const
{
    QVector<Ability*>* counters = this->counters;
    return counters == nullptr ? 0 : counters->size();
}

int Costume::getElementResistance(const int element) const
{
    QMap<int, int>* res = this->res;
    return res == nullptr ? 0 : res->value(element, 0);
}

int Costume::getStateResistance(State* const state) const
{
    QMap<State*, int>* stRes = this->stRes;
    return stRes == nullptr ? 0 : stRes->value(state, 0);
}

bool Costume::isReflecting() const
{
    return (this->flags & FLAG_REFLECT) == FLAG_REFLECT;
}

bool Costume::isCountering() const
{
    QVector<Ability*>* counters = this->counters;
    return counters != nullptr && counters->size() > 0;
}

bool Costume::isShapeShifted() const
{
    return (this->flags & FLAG_SHAPE_SHIFT) == FLAG_SHAPE_SHIFT;
}

bool Costume::isEnraged() const
{
    return (this->flags & FLAG_ENRAGED) == FLAG_ENRAGED;
}

bool Costume::isConfused() const
{
    return (this->flags & FLAG_CONFUSE) == FLAG_CONFUSE;
}

bool Costume::isConverted() const
{
    return (this->flags & FLAG_CONVERT) == FLAG_CONVERT;
}

bool Costume::isInvincible() const
{
    return (this->flags & FLAG_INVINCIBLE) == FLAG_INVINCIBLE;
}

bool Costume::isStunned() const
{
    return (this->flags & FLAG_STUN) == FLAG_STUN;
}

bool Costume::isKnockedOut() const
{
    return (this->flags & FLAG_KO) == FLAG_KO;
}

Costume& Costume::adopt(QString& ret, Actor& actor)
{
    return this->adopt(&ret, nullptr, actor, true, false);
}

Costume& Costume::abandon(QString& ret, Actor& actor)
{
    return this->adopt(&ret, nullptr, actor, true, true);
}

Costume& Costume::apply(QString& ret, Actor& actor)
{
    return this->apply(ret, nullptr, actor);
}

Costume& Costume::adopt(QString* const ret, Scene* const scene, Actor& actor, bool const updStates, bool const remove)
{
    Costume& costume = *this;
    actor.updateAttributes(remove, scene, costume);
    actor.updateResistance(remove, costume.res, costume.stRes);
    if (remove)
    {
        int const roleFlags = costume.flags;
        int const actorFlags = actor.flags;
        if ((actorFlags & roleFlags) == roleFlags)
        {
            actor.flags = actorFlags ^ roleFlags;
        }
        if (costume.isShapeShifted() && costume.sprite != nullptr)
        {
            (*actor.sprite) = *(actor.getJob().sprite);
        }
        if (costume.hp == 0 && costume.mp == 0 && costume.sp == 0)
        {
            int const roleElm = costume.dmgType;
            int const actorElm = actor.dmgType;
            if ((actorElm & roleElm) == roleElm)
            {
                actor.dmgType = actorElm ^ roleElm;
            }
        }
        else
        {
            QVector<Costume*>* const dmgRoles = actor.dmgRoles;
            if (dmgRoles != nullptr)
            {
                dmgRoles->removeOne(&costume);
            }
        }
        costume.refresh(ret, scene, actor, updStates, true);
        actor.refreshCostumes(ret, scene);
    }
    else
    {
        if (updStates)
        {
            QMap<State*, int>* const cStates = costume.stateDur;
            if (cStates != nullptr)
            {
                actor.updateStates(false, ret, scene, *cStates, true);
            }
        }
        if (costume.hp != 0 || costume.mp != 0 || costume.sp != 0)
        {
            QVector<Costume*>* dmgRoles = actor.dmgRoles;
            if (dmgRoles == nullptr)
            {
                dmgRoles = new QVector<Costume*>();
                actor.dmgRoles = dmgRoles;
            }
            dmgRoles->append(&costume);
        }
        costume.refresh(ret, scene, actor, false, false);
    }
    return costume;
}

Costume& Costume::apply(QString& ret, Scene* scene, Actor& actor)
{
    Costume& role = *this;
    ret = ret % QString(Costume::CausesTxt).arg(actor.name, role.name);
    role.damage(ret, scene, nullptr, actor, std::rand() % 4, true);
    return role;
}

Costume& Costume::refresh(QString* const ret, Scene* const scene, Actor& actor, bool const updStates, bool const remove)
{
    Costume& costume = *this;
    {
        QVector<Ability*>* skills = costume.aSkills;
        if (skills != nullptr)
        {
            actor.updateSkills(remove, false, *skills);
        }
        skills = costume.counters;
        if (skills != nullptr)
        {
            actor.updateSkills(remove, true, *skills);
        }
    }
    if (updStates)
    {
        QMap<State*, int>* cStates = costume.stateDur;
        if (cStates != nullptr)
        {
            actor.updateStates(remove, ret, scene, *cStates, false);
        }
    }
    if (!remove)
    {
        int const cFlags = costume.flags;
        if (costume.hp == 0 && costume.mp == 0 && costume.sp == 0)
        {
            actor.dmgType |= costume.dmgType;
        }
        actor.flags |= cFlags;
        if (scene != nullptr && ret != nullptr && (cFlags & FLAG_KO) == FLAG_KO)
        {
            scene->checkStatus(*ret);
        }
        if (costume.isShapeShifted())
        {
            QString* const spr = costume.sprite;
            if (spr != nullptr)
            {
                (*actor.sprite) = *spr;
            }
        }
    }
    return actor;
}

Costume::Costume(int const id, QString& name, QString& sprite, bool const shapeShift, int const mActions, int const elm, int const hpDmg, int const mpDmg, int const spDmg,
                 int const mHp, int const mMp, int const mSp, int const atk, int const def, int const spi, int const wis, int const agi, bool const stun, bool const range,
                 bool const automate, bool const confuse, bool const reflect, bool const ko, bool const invincible, bool const revive, QVector<Ability*>* const skills,
                 QVector<Ability*>* const counters, QMap<State*, int>* const states, QMap<State*, int>* const stRes, QMap<int, int>* const res)
    : Role(id, name, sprite, hpDmg, mpDmg, spDmg, mHp, mMp, mSp, elm, range, revive, states)
{
    this->atk = atk;
    this->def = def;
    this->spi = spi;
    this->wis = wis;
    this->agi = agi;
    this->res = res;
    this->mActions = mActions;
    this->stRes = stRes;
    this->aSkills = skills;
    this->counters = counters;
    int flags = this->flags;
    if (stun)
    {
        flags |= FLAG_STUN;
    }
    if (shapeShift)
    {
        QString* spr = this->sprite;
        if (spr != nullptr && spr->length() > 0)
        {
            flags |= FLAG_SHAPE_SHIFT;
        }
    }
    if (automate)
    {
        flags |= FLAG_ENRAGED;
    }
    if (confuse)
    {
        flags |= FLAG_CONFUSE;
    }
    if (reflect)
    {
        flags |= FLAG_REFLECT;
    }
    if (invincible)
    {
        flags |= FLAG_INVINCIBLE;
    }
    if (ko)
    {
        flags |= FLAG_KO;
    }
    this->flags = flags;
}

Costume::Costume(int const id, QString name, QString sprite, bool const shapeShift, int const mActions, int const elm, int const hpDmg, int const mpDmg, int const spDmg,
                 int const mHp, int const mMp, int const mSp, int const atk, int const def, int const spi, int const wis, int const agi, bool const range, bool const automate,
                 bool const confuse, bool const reflect, bool const invincible, bool const revive, QVector<Ability*>* const skills, QVector<Ability*>* const counters,
                 QMap<State*, int>* const states, QMap<State*, int>* const stRes, QMap<int, int>* const res)
    : Costume(id, name, sprite, shapeShift, mActions, elm, hpDmg, mpDmg, spDmg, mHp, mMp, mSp, atk, def, spi, wis, agi, false, range, automate, confuse, reflect, false, invincible,
              revive, skills, counters, states, stRes, res)
{

}

Costume::Costume(Costume& costume) : Role(costume)
{
    this->atk = costume.atk;
    this->def = costume.def;
    this->spi = costume.spi;
    this->wis = costume.wis;
    this->agi = costume.agi;
    this->res = costume.res;
    this->aSkills = costume.aSkills;
    this->mActions = costume.mActions;
    this->stRes = costume.stRes;
}

Costume::~Costume()
{

}

File: itbrpgsca/itbrpgsca/Costume.swift
/*
Copyright (C) AD 2013-2020 Claudiu-Stefan Costea

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
*/

import Foundation

open class Costume : Role {
    
    public static let FLAG_ENRAGED = 4
    public static let FLAG_CONFUSE = 8
    public static let FLAG_CONVERT = 16
    public static let FLAG_SHAPE_SHIFT = 32
    public static let FLAG_INVINCIBLE = 64
    public static let FLAG_COVER = 128
    public static let FLAG_DRAW = 256
    public static let FLAG_STUN = 512
    public static let FLAG_KO = 1024
    
    public static var CausesTxt = ", %@ is affected by %@"
    
    internal var _atk: Int, _def: Int, _spi: Int, _wis: Int, _agi: Int, _mActions: Int, _rflType: Int,
                 _aSkills: [Ability]?, _counters: [Ability]?, _stRes: [State?: Int]?, _res: [Int: Int]?
    
    open var enraged: Bool {
        return (self._flags & Costume.FLAG_ENRAGED) == Costume.FLAG_ENRAGED
    }
    
    open var confused: Bool {
        return (self._flags & Costume.FLAG_CONFUSE) == Costume.FLAG_CONFUSE
    }
    
    open var converted: Bool {
        return (self._flags & Costume.FLAG_CONVERT) == Costume.FLAG_CONVERT
    }
    
    open var invincible: Bool {
        return (self._flags & Costume.FLAG_INVINCIBLE) == Costume.FLAG_INVINCIBLE
    }
    
    open var shapeShifted: Bool {
        return (self._flags & Costume.FLAG_SHAPE_SHIFT) == Costume.FLAG_SHAPE_SHIFT
    }
    
    open var covered: Bool {
        return (self._flags & Costume.FLAG_COVER) == Costume.FLAG_COVER
    }
    
    open var drawn: Bool {
        return (self._flags & Costume.FLAG_DRAW) == Costume.FLAG_DRAW
    }
    
    open var stunned: Bool {
        return (self._flags & Costume.FLAG_STUN) == Costume.FLAG_STUN
    }
    
    open var knockedOut: Bool {
        return (self._flags & Costume.FLAG_KO) == Costume.FLAG_KO
    }
    
    open var reflects: Bool {
        return self._rflType != 0
    }
    
    open var counters: [Ability]? {
        return self._counters
    }
    
    open var stRes: [State?: Int]? {
        return self._stRes
    }
    
    open var aSkills: [Ability]? {
        return self._aSkills
    }
    
    open var res: [Int: Int]? {
        return self._res
    }
    
    open var rflType: Int {
        return self._rflType
    }
    
    open var atk: Int {
        return self._atk
    }
    
    open var def: Int {
        return self._def
    }
    
    open var spi: Int {
        return self._spi
    }
    
    open var wis: Int {
        return self._wis
    }
    
    open var agi: Int {
        return self._agi
    }
    
    open var mActions: Int {
        return self._mActions
    }
    
    open func refresh(actor: Actor, updStates: Bool, remove: Bool) {
        if let skills = self.aSkills {
            actor.updateSkills(remove: remove, counters: false, skills: skills)
        }
        if let counters = self.counters {
            actor.updateSkills(remove: remove, counters: true, skills: counters)
        }
        if updStates, let cStates = self.stateDur {
            actor.updateStates(remove: true, states: cStates, withDur: false)
        }
        if remove {
            let roleFlags = self._flags
            let actorFlags = actor.flags
            if (actorFlags & roleFlags) == roleFlags {
                actor.flags = actorFlags ^ roleFlags
            }
            if self.shapeShifted, self.sprite != nil {
                actor.sprite = actor.job.sprite
            }
            let roleRfl = self.rflType
            let actorRfl = actor.rflType
            if (actorRfl & roleRfl) == roleRfl {
                actor.rflType = actorRfl & rflType
            }
            if self.hp == 0 && self.mp == 0 && self.sp == 0 {
                let roleElm = self.dmgType
                let actorElm = self.dmgType
                if (actorElm & roleElm) == roleElm {
                    actor.dmgType = actorElm ^ roleElm
                }
            } else if var dmgRoles = actor._dmgRoles {
                dmgRoles.removeAll(where: { $0 == self })
            }
            actor.refreshCostumes()
        } else {
            actor.flags |= self._flags
            if self.hp == 0 && self.mp == 0 && self.sp == 0 {
                actor.dmgType |= self.dmgType
            }
            actor.rflType |= self.rflType
            if self.shapeShifted, let spr = self.sprite {
                actor.sprite = spr
            }
        }
    }
    
    open func adopt(actor: Actor, addStates: Bool) {
        self.adopt(actor: actor, updStates: addStates, remove: false)
    }
    
    open func adopt(actor: Actor, updStates: Bool, remove: Bool) {
        actor.updateAttributes(remove: remove, costume: self)
        actor.updateResistance(remove: remove, elmRes: self.res, stRes: self.stRes)
        if remove {
            self.refresh(actor: actor, updStates: updStates, remove: true)
        } else {
            if updStates {
                if let cStates = self.stateDur {
                    actor.updateStates(remove: false, states: cStates, withDur: true)
                }
            }
            if self.hp != 0 || self.mp != 0 || self.sp != 0 {
                var dmgRoles: [Costume]! = actor._dmgRoles
                if dmgRoles == nil {
                    dmgRoles = [Costume]()
                    actor._dmgRoles = dmgRoles
                }
                dmgRoles.append(self)
            }
            self.refresh(actor: actor, updStates: false, remove: false)
        }
    }
    
    open func apply(ret: inout String, actor: Actor) {
        ret.append(String(format: Costume.CausesTxt, actor.name, self.name))
        self.damage(ret: &ret, absorber: nil, actor: actor, dmg: Int.random(in: 0...4), percent: true)
    }
    
    open func abandon(actor: Actor, delStates: Bool) {
        self.adopt(actor: actor, updStates: delStates, remove: true)
    }
    
    init(id: Int, name: String, sprite: String?, shapeShift: Bool, mActions: Int, mDelayTrn: Int, dmgType: Int, rflType: Int,
         hpDmg: Int, mpDmg: Int, spDmg: Int, mHp: Int, mMp: Int, mSp: Int, atk: Int, def: Int, spi: Int, wis: Int, agi: Int,
         stun: Bool, range: Bool, enrage: Bool, confuse: Bool, invincible: Bool, ko: Bool, revive: Bool, skills: [Ability]?,
         counters: [Ability]?, states: [State: Int]?, stRes: [State: Int]?, res: [Int: Int]?) {
        self._counters = counters
        self._aSkills = skills
        self._stRes = stRes
        self._res = res
        self._atk = atk
        self._def = def
        self._spi = spi
        self._wis = wis
        self._agi = agi
        self._rflType = rflType
        self._mActions = mActions
        super.init(id: id, name: name, sprite: sprite, hp: hpDmg, mp: mpDmg, sp: spDmg, mHp: mHp, mMp: mMp, mSp: mSp,
                   mDelayTrn: mDelayTrn, dmgType: dmgType, range: range, revive: revive, stateDur: states)
        var flags = self._flags
        if shapeShift {
            flags |= Costume.FLAG_SHAPE_SHIFT
        }
        if invincible {
            flags |= Costume.FLAG_INVINCIBLE
        }
        if enrage {
            flags |= Costume.FLAG_ENRAGED
        }
        if confuse {
            flags |= Costume.FLAG_CONFUSE
        }
        if stun {
            flags |= Costume.FLAG_STUN
        }
        if ko {
            flags |= Costume.FLAG_KO
        }
        self._flags = flags
    }
    
}
